trin-core/src/portalnet/find/query_trace.rs

```diff
@@ -0,0 +1,219 @@
+            node_id_string,
+            QueryResponse {
+                timestamp_millis: timestamp_u64,
+                responded_with: vec![],
+            },
+        );
+        self.add_node_metadata(enr, true);
+    }
+
+    /// Returns milliseconds since the time provided.
+    fn timestamp_millis_u64(since: SystemTime) -> u64 {
+        let timestamp_millis_u128 = SystemTime::now()
+            .duration_since(since)
+            .unwrap_or(Duration::default())
+            .as_millis();
+        // JSON serialization does not support u128. u64 can hold a few million years worth of milliseconds.
+        u64::try_from(timestamp_millis_u128).unwrap_or(u64::MAX)
+    }
+
+    fn add_node_metadata(&mut self, enr: &Enr, node_responded: bool) {
+        let target = self.target_id;
+        self.node_metadata
+            .entry(hex_encode(enr.node_id().raw()))
+            .and_modify(|node_info| {
+                // If we see a different ENR for a Node ID we've already seen,
+                // use the newer ENR if its seq # is higher. If the seq #s are equal
+                // use the ENR that we received a response from.
+                let different_enr: bool = *enr != node_info.enr;
+                let higher_incoming_seq_number: bool = enr.seq() > node_info.enr.seq();
+                let same_incoming_seq_number: bool = enr.seq() == node_info.enr.seq();
+                if different_enr
+                    && ((same_incoming_seq_number && node_responded) || higher_incoming_seq_number)
+                {
+                    *node_info = QueryTrace::get_node_metadata(enr, &target);
+                }
+            })
+            .or_insert_with(|| -> NodeInfo { QueryTrace::get_node_metadata(enr, &target) });
+    }
+
+    fn get_node_metadata(enr: &Enr, target: &ContentId) -> NodeInfo {
+        let node_id = enr.node_id();
+        let node_id_raw = node_id.raw();
+        let ip = match enr.ip4() {
+            Some(ip) => ip.to_string(),
+            None => "".to_owned(),
+        };
+        let port = enr.udp4().unwrap_or(0).to_string();
+        let distance = XorMetric::distance(&node_id_raw, &target.raw());
+        let distance_log2 = distance.log2().unwrap_or(0);
+        let distance: u64 = distance.0[3];
+        let distance = (distance >> 32) as u32;
+
+        NodeInfo {
+            enr: enr.clone(),
+            ip,
+            port,
+            distance_to_content: distance,
+            distance_log2,
+        }
+    }
+}
+
+/// Represents the response from a single node.
+#[derive(Debug, Serialize)]
+pub struct QueryResponse {
+    /// Milliseconds since query started.
+    pub timestamp_millis: u64,
+    /// Node IDs that this node responded with.
+    pub responded_with: Vec<NodeIdString>,
+}
+
+#[derive(Debug, Serialize)]
+pub struct NodeInfo {
+    pub enr: Enr,
+    pub ip: String,
+    pub port: String,
+    pub distance_to_content: u32,
+    pub distance_log2: usize,
+}
```

Comment:
From above:

> Note that node_meta_data was added to the format, but will no longer be necessary once full ENR parsing is working on the front-end.

There are some issues with ENR parsing on the front-end that I haven't sorted out. 

u32 distance and log2 distance so that we can sort nodes with the same log2 distance by distance. I could instead calculate log2 of distance on the frontend.